packages/core/crates/core-crypto/src/derivation.rs

```diff
@@ -0,0 +1,147 @@
+use crate::primitives::calculate_mac;
+use crate::random::random_fill;
+
+// Module-specific constants
+const HASH_KEY_COMMITMENT_SEED: &str = "HASH_KEY_COMMITMENT_SEED";
+const HASH_KEY_HMAC: &str = "HASH_KEY_HMAC";
+const HASH_KEY_PACKET_TAG: &str = "HASH_KEY_PACKET_TAG";
+
+/// Helper function to expand an already cryptographically strong key material using the HKDF expand function
+fn hkdf_expand_from_prk<const OUT_LENGTH: usize>(secret: &[u8], tag: &[u8]) -> Result<[u8; OUT_LENGTH]> {
+    // Create HKDF instance
+    let hkdf = SimpleHkdf::<Blake2s256>::from_prk(secret)
+        .map_err(|_| InvalidParameterSize{name: "secret".into(), expected: SECRET_KEY_LENGTH})?;
+
+    // Expand the key to the required length
+    let mut out = [0u8; OUT_LENGTH];
+    hkdf.expand(tag, &mut out)
+        .map_err(|_| InvalidInputValue)?;
+
+    Ok(out)
+}
+
+/// Derives a ping challenge (if no challenge is given) or a pong response to a ping challenge.
+pub fn derive_ping_pong(challenge: Option<&[u8]>) -> Box<[u8]> {
+    let mut ret = [0u8; PING_PONG_NONCE_SIZE];
+    match challenge {
+       None => random_fill(&mut ret),
+       Some(chal) => {
+           let mut digest = Blake2s256::default();
+           digest.update(chal);
+           // Finalize requires enough space for the hash value, so this needs an extra copy
+           let hash = digest.finalize_fixed().to_vec();
+           ret.copy_from_slice(&hash[0..PING_PONG_NONCE_SIZE]);
+       }
+    }
+    ret.into()
+}
+
+/// Derives the commitment seed given the compressed private key representation
+/// and the serialized channel information.
+pub fn derive_commitment_seed(private_key: &[u8], channel_info: &[u8]) -> Result<Box<[u8]>> {
+    hkdf_expand_from_prk::<SECRET_KEY_LENGTH>(private_key, HASH_KEY_COMMITMENT_SEED.as_bytes())
+        .and_then(|key| calculate_mac(&key, channel_info))
+}
+
+/// Derives the packet tag used during packet construction by expanding the given secret.
+pub fn derive_packet_tag(secret: &[u8]) -> Result<Box<[u8]>> {
+    hkdf_expand_from_prk::<PACKET_TAG_LENGTH>(secret, HASH_KEY_PACKET_TAG.as_bytes())
+        .map(Box::from)
+}
+
+/// Derives a key for MAC calculation by expanding the given secret.
+pub fn derive_mac_key(secret: &[u8]) -> Result<Box<[u8]>> {
+    hkdf_expand_from_prk::<SECRET_KEY_LENGTH>(secret, HASH_KEY_HMAC.as_bytes())
+        .map(Box::from)
+}
+
+pub(crate) fn generate_key_iv(secret: &[u8], info: &[u8], key: &mut [u8], iv: &mut [u8], iv_first: bool) -> Result<()> {
+    if secret.len() != SECRET_KEY_LENGTH {
+        return Err(InvalidParameterSize{name: "secret".into(), expected: SECRET_KEY_LENGTH})
+    }
+
+    let hkdf = SimpleHkdf::<Blake2s256>::from_prk(secret)
+        .map_err(|_| InvalidParameterSize{name: "secret".into(), expected: SECRET_KEY_LENGTH})?;
+
+    let mut out = vec![0u8; key.len() + iv.len()];
+    hkdf.expand(info, &mut out)
+        .map_err(|_| InvalidInputValue)?;
+
+    if iv_first {
```

Comment:
Why do we need this? Looks like this easily allows misusage

Reply:
I agree, but PRP and PRG sample key and IV in opposite order from the keying material:
See:
- https://github.com/hoprnet/hoprnet/blob/31fb0aee462d1c758c3f8b7dc9dc64b8effbc5f3/packages/core/crates/core-crypto/src/prg.rs#L42
- https://github.com/hoprnet/hoprnet/blob/31fb0aee462d1c758c3f8b7dc9dc64b8effbc5f3/packages/core/crates/core-crypto/src/prp.rs#L41